app/celer-geo/Runner.cc

```diff
@@ -0,0 +1,228 @@
+ * Perform a raytrace.
+ */
+auto Runner::operator()(TraceSetup const& trace, ImageInput const& image_inp)
+    -> SPImage
+{
+    // Create image params
+    last_image_ = std::make_shared<ImageParams>(image_inp);
+
+    return (*this)(trace);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Perform a raytrace using the last image but a new geometry/memspace.
+ */
+auto Runner::operator()(TraceSetup const& trace) -> SPImage
+{
+    CELER_VALIDATE(last_image_,
+                   << "first trace input did not specify an image");
+
+    // Load geometry
+    SPImager imager = this->make_imager(trace.geometry);
+
+    // Create image
+    SPImage image = this->make_traced_image(trace.memspace, *imager);
+    return image;
+}
+//---------------------------------------------------------------------------//
+/*!
+ * Get volume names from an already loaded geometry.
+ */
+std::vector<std::string> Runner::get_volumes(Geometry g) const&
+{
+    CELER_EXPECT(geo_cache_[g]);
+
+    auto const& geo = *geo_cache_[g];
+    std::vector<std::string> result(geo.num_volumes());
+    for (auto i : range<VolumeId::size_type>(result.size()))
+    {
+        result[i] = geo.id_to_label(VolumeId{i}).name;
+    }
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Load a geometry, caching it.
+ *
+ * If Geant4 is available and the input file is GDML, this will be executed
+ * when the runner is constructed to save a reusable pointer to the Geant4
+ * world. Otherwise, this is called by the imager when raytracing a new
+ * geometry type.
+ */
+template<Geometry G>
+auto Runner::load_geometry() -> std::shared_ptr<GeoParams_t<G> const>
+{
+    using GP = GeoParams_t<G>;
+
+    auto& cached = geo_cache_[G];
+    std::shared_ptr<GP const> geo;
+    if constexpr (!is_geometry_configured_v<GP>)
+    {
+        CELER_NOT_CONFIGURED(to_cstring(G));
+    }
+    else if (cached)
+    {
+        // Downcast to type
+        geo = std::dynamic_pointer_cast<GP const>(cached);
+    }
+    else
+    {
+        Stopwatch get_time;
+        if (geant_world_)
+        {
+            // Load from existing Geant4 geometry
+            geo = std::make_shared<GP>(geant_world_);
+        }
+        else
+        {
+            // Load directly from input file
+            geo = std::make_shared<GP>(input_.geometry_file);
+            if constexpr (G == Geometry::geant4)
+            {
+                // Save world for later reuse
+                geant_world_ = static_cast<GP const&>(*geo).world();
+            }
+        }
+        // Save load time
+        timers_[std::string{"load_"} + to_cstring(G)] = get_time();
+
+        // Save geometry in cache
+        cached = geo;
+    }
+
+    CELER_ENSURE(cached);
+    CELER_ENSURE(geo);
+    return geo;
+}
+
+//---------------------------------------------------------------------------//
+#define CASE_RETURN_FUNC_T(T, FUNC, ...) \
```

Comment:
Not necessary defining macros near why there are used, but confirm this placement was intention since it looks pretty strange where it is.

Reply:
Yeah, this is definitely a weird place. I would've put it inside the function except I used it in two functions. I think it would look better with a namespace and put at the top of the file.